src/coreclr/jit/switchrecognition.cpp

```diff
@@ -132,6 +139,204 @@ bool IsConstantTestCondBlock(const BasicBlock* block,
+
+    // In most cases the switch value is either ADD(LCL_VAR, CNS) or LCL_VAR
+    if (switchOp->OperIs(GT_ADD))
+    {
+        // We use gtEffectiveVal() to skip COMMA nodes, those will be evaluated before
+        // any other case node in the switch block anyway.
+        GenTree* op1 = switchOp->gtGetOp1()->gtEffectiveVal();
+        GenTree* op2 = switchOp->gtGetOp2();
+
+        // TODO-JumpTable: We should be able to accept any side-effect-free op1 here
+        if (op1->OperIs(GT_LCL_VAR) && op2->IsCnsIntOrI())
+        {
+            *variableNode = op1;
+            *offset       = op2->AsIntCon()->IconValue();
+            return true;
+        }
+    }
+    else
+    {
+        // We use gtEffectiveVal() to skip COMMA nodes, those will be evaluated before
+        // any other case node in the switch block anyway.
+        GenTree* op1 = switchOp->gtEffectiveVal();
+        if (op1->OperIs(GT_LCL_VAR))
+        {
+            *variableNode = op1;
+            *offset       = 0;
+            return true;
+        }
+    }
+    return false;
+}
+
+//------------------------------------------------------------------------------
+// optExtendSwitch: If one of the switch's targets is a value test block,
+//    and the switch's variable is the same as the value test's variable - try to merge
+//    that test (BBJ_COND) into the switch block. It's useful to assist Roslyn, which
+//    uses fairly conservative heuristics to generate switch blocks and may give up on
+//    some tests that could be merged into the switch.
+//
+// Arguments:
+//    block - BBJ_SWITCH block to extend
+//
+// Return Value:
+//    True if switch was successfully extended, false otherwise
+//
+bool Compiler::optExtendSwitch(BasicBlock* block)
+{
+    assert(block->KindIs(BBJ_SWITCH));
+    JITDUMP("Considering expanding switch " FMT_BB "\n", block->bbNum);
+
+    GenTree* switchOp           = nullptr;
+    ssize_t  switchTargetOffset = 0;
+    if (!GetSwitchValueOp(block, &switchOp, &switchTargetOffset))
+    {
+        JITDUMP("Couldn't extract the switch's variable - bail out.\n");
+        return false;
+    }
+
+    // TODO-JumpTable: Currently, we only expand the default target, but it should be possible
+    // to traverse all non-default edges as well (if they point to similar value tests).
+    // It shouldn't be too hard to implement (might be TP costly, though).
+
+    BBswtDesc* switchTargets = block->GetSwitchTargets();
+    if (!switchTargets->bbsHasDefault)
+    {
+        JITDUMP("Switch doesn't have a default target - bail out.\n");
+        return false;
+    }
+
+    // Now let's see if we can merge the value test in the edgeToExpand's destination block
+    // with the switch block.
+    FlowEdge*   edgeToExpand    = switchTargets->getDefault();
+    BasicBlock* edgeToExpandDst = edgeToExpand->getDestinationBlock();
+
+    // IsConstantTestCondBlock will tell us if the default target is a value test block (e.g. "x == cns")
```

Comment:
I think you also need to check if `edgeToExpandDst` has other statements, and if so, bail out.

Reply:
Ah, oops, good point! I thought that `IsConstantTestCondBlock` checked for that but it asks caller for that

Reply:
Addressed